bier/struct/BinarySerializable.py

```diff
@@ -103,72 +71,104 @@ def write_to(self, writer):
+        if get_origin_type(base) is BinarySerializable:
+            return base
+
+    raise ValueError(f"Type {cls} does not inherit from BinarySerializable")
+
+
+def parse_length_type(annotation: Annotated) -> TypeNode[int]:
     args = get_args(annotation)
-    assert len(args) >= 2, "Annotated length must have two arguments"
-    assert args[0] is int, "Annotated length must be an int"
+    assert len(args) >= 2, "Length type must have two arguments"
+    assert args[0] is int, "Length type must be an int"
     if issubclass(args[1], TypeNode):
         return args[1]()
     elif isinstance(args[1], TypeNode):
         return args[1]
     else:
         raise ValueError(
-            f"Annotated length must be a TypeNode[int] or an instance of it, got {args[1]} instead."
+            f"Length type must be a TypeNode[int] or an instance of it, got {args[1]} instead."
         )
 
 
-def parse_annotation(annotation: type, options: BinarySerializableOptions) -> TypeNode:
-    if isinstance(annotation, str):
-        raise ValueError(
-            f"Annotation is a string: {annotation}. Use resolve_annotation_str to resolve it."
-        )
-
+def parse_annotation(annotation: Any, options: BinarySerializableOptions) -> TypeNode:
     if annotation in PRIMITIVES:
         args = get_args(annotation)
         assert len(args) >= 2 and issubclass(args[1], TypeNode), (
             f"Invalid primitive {annotation}"
         )
         return args[1]()
 
-    origin = get_origin(annotation) or annotation
+    origin = get_origin_type(annotation)
     args = get_args(annotation)
 
     if annotation is cstr:
         return StringNode()
 
     if annotation is str:
-        return StringNode(options.default_length_encoding)
+        return StringNode(options.length_type)
 
     if origin is list:
         assert len(args) == 1, "list must have one argument"
         elem_node = parse_annotation(
             annotation=args[0],
             options=options,
         )
-        return ListNode(elem_node, options.default_length_encoding)
-
-    if origin is list_d:
-        assert len(args) == 2, "list_d must have two arguments"
-        raw_elem, raw_size = args
-        elem_node = parse_annotation(
-            annotation=raw_elem,
-            options=options,
-        )
-        size_node = parse_annotated_length(annotation=raw_size)
-        return ListNode(elem_node, size_node)
+        return ListNode(elem_node, options.length_type)
 
     if origin is tuple:
         return TupleNode(
             tuple(parse_annotation(annotation=arg, options=options) for arg in args)
         )
 
-    if origin is str_d:
-        assert len(args) == 1, "str_d must have one argument"
-        size_node = parse_annotated_length(annotation=args[0])
-        return StringNode(size_node)
+    if origin is custom:
+        assert len(args) >= 1, "member must have at least one argument"
+        member_type = args[0]
+        member_options = replace(options)
```

Comment:
That doesn't seem like a good usage of replace?

Might be better to parse the options into a new dict,
and then "apply" it via replace